libs/fujitsu-thermal-printer/src/printing.ts

```diff
@@ -0,0 +1,351 @@
+  }
+
+  while (i < data.length) {
+    const byte = data[i] as number;
+    // if a lone final byte, encode as literal
+    if (i + 1 >= data.length) {
+      literalBuffer.push(byte);
+      flushLiteralBuffer();
+      break;
+    }
+
+    const nextByte = data[i + 1] as number;
+
+    if (byte === nextByte) {
+      flushLiteralBuffer();
+
+      // encode repeating bytes
+      let repeats = 2;
+      while (
+        i + repeats < data.length &&
+        repeats < MAX_PACKET_DATA_LENGTH &&
+        data[i + repeats] === byte
+      ) {
+        repeats += 1;
+      }
+      compressedData.push(1 - repeats, byte);
+      i += repeats;
+    } else {
+      literalBuffer.push(byte);
+      if (literalBuffer.length === MAX_PACKET_DATA_LENGTH) {
+        flushLiteralBuffer();
+      }
+      i += 1;
+    }
+  }
+
+  flushLiteralBuffer();
+
+  return new Int8Array(compressedData);
+}
+
+export function compressBitImage(
+  uncompressed: UncompressedBitImage
+): CompressedBitImage {
+  debug('compressing bit image');
+  return {
+    height: uncompressed.height,
+    data: packBitsCompression(uncompressed.data),
+    compressed: true,
+  };
+}
+
+const WAIT_FOR_BUFFER_NOT_FULL_TIMEOUT_MS = 2.5 * 1000;
+const WAIT_FOR_BUFFER_FLUSH_TIMEOUT_MS = 10 * 1000;
+
+export async function printPageBitImage(
+  driver: FujitsuThermalPrinterDriverInterface,
+  compressedBitImages: IteratorPlus<CompressedBitImage>
+): Promise<Result<void, RawPrinterStatus>> {
+  // print all bit images that compose the page
+  await driver.setReplyParameter(PRINT_ONGOING_REPLY_PARAMETER);
+  for (const compressedBitImage of compressedBitImages) {
+    // wait for the buffer to be ready before sending another bit image. this
+    // is not strictly necessary - either the OS or `node-usb` layer handles
+    // buffering - but it allows us to keep our execution roughly in sync with
+    // the hardware and allows us to bail earlier if the printer stops
+    const waitForPrintReadyResult = await waitForPrintReadyStatus(driver, {
+      interval: 100,
+      timeout: WAIT_FOR_BUFFER_NOT_FULL_TIMEOUT_MS,
+      replyParameter: PRINT_ONGOING_REPLY_PARAMETER,
+    });
+    if (waitForPrintReadyResult.isErr()) {
+      return waitForPrintReadyResult;
+    }
+
+    assert(compressedBitImage);
+    driver.printBitImage(compressedBitImage);
+  }
+
+  await driver.setReplyParameter(PRINT_PROCESSING_REPLY_PARAMETER);
+  const waitForPrintFinished = await waitForPrintReadyStatus(driver, {
+    interval: 100,
+    timeout: WAIT_FOR_BUFFER_FLUSH_TIMEOUT_MS,
+    replyParameter: PRINT_PROCESSING_REPLY_PARAMETER,
+  });
+  if (waitForPrintFinished.isErr()) {
+    return waitForPrintFinished;
+  }
+
+  debug('printed page successfully');
+  return ok();
+}
+
+export async function print(
+  driver: FujitsuThermalPrinterDriverInterface,
+  pdfData: Uint8Array
+): Promise<Result<void, RawPrinterStatus>> {
+  const pdfImages = pdfToImages(Buffer.from(pdfData), {
+    scale: 200 / 72,
```

Comment:
I think I can guess these numbers are, but could we add comments or constant definitions for clarity?